src/Sentry/Internal/Http/CachingTransport.cs

```diff
@@ -0,0 +1,266 @@
+                    {
+                        _options.DiagnosticLogger?.LogError(
+                            "Transient failure when sending cached envelope: {0}",
+                            ex,
+                            envelope.TryGetEventId()
+                        );
+
+                        shouldRetry = true;
+                    }
+                    catch (HttpRequestException ex)
+                    {
+                        _options.DiagnosticLogger?.LogError(
+                            "Persistent failure when sending cached envelope: {0}",
+                            ex,
+                            envelope.TryGetEventId()
+                        );
+                    }
+                }
+
+                // Envelope & file stream must be disposed prior to reaching this point
+
+                if (shouldRetry)
+                {
+                    // Move the file back to cache directory so it gets processed again in the future
+                    Directory.CreateDirectory(_cacheDirectoryPath);
+
+                    File.Move(
+                        envelopeFilePath,
+                        Path.Combine(_cacheDirectoryPath, Path.GetFileName(envelopeFilePath))
+                    );
+
+                    // Avoid processing other files because the order may end up corrupted
+                    break;
+                }
+                else
+                {
+                    // Delete the envelope file and move on to the next one
+                    File.Delete(envelopeFilePath);
+                }
+            }
+        }
+
+        private async ValueTask StoreToCacheAsync(
+            Envelope envelope,
+            CancellationToken cancellationToken = default)
+        {
+            using var lockClaim = await _cacheDirectoryLock.ClaimAsync(cancellationToken).ConfigureAwait(false);
+
+            // If over capacity - remove oldest envelope file
+            while (
+                GetCacheFilePaths().Count() >= _options.MaxQueueItems &&
+                GetCacheFilePaths().FirstOrDefault() is { } oldestEnvelopeFilePath)
+            {
+                File.Delete(oldestEnvelopeFilePath);
+            }
+
+            // Envelope file name can be either:
+            // 1604679692_b2495755f67e4bb8a75504e5ce91d6c1_17754019.envelope
+            // 1604679692__17754019.envelope
+            // (depending on whether event ID is present or not)
+            var envelopeFilePath = Path.Combine(
+                _cacheDirectoryPath,
+                $"{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}_" +
+                $"{envelope.TryGetEventId()}_" +
+                $"{envelope.GetHashCode()}" +
+                $".{EnvelopeFileExt}"
+            );
+
+            Directory.CreateDirectory(_cacheDirectoryPath);
+
+            using (var stream = File.Create(envelopeFilePath))
+            {
+                await envelope.SerializeAsync(stream, cancellationToken).ConfigureAwait(false);
```

Comment:
If an IOException like Disk full happens, it'll not release the workerSignal?

Reply:
No, it won't.